sympy/solvers/ode.py

```diff
@@ -1044,6 +1037,47 @@ class in it.  Note that a hint may do this anyway if
             else:
                 raise ValueError("Enter boundary conditions of the form ics={f(point}: value, f(x).diff(x, order).subs(x, point): value}")
 
+
+    # Precondition to try remove unnecessary factor and power from equation
+    eqs = eq
+    if isinstance(eqs, Pow):
+        eqs_args = eqs.args # if f(x)**p=0 then f(x)=0 (p>0)
+        if eqs_args[1]>0:
+            eqs = eqs_args[0]
+
+    # I will try not to change the structure of equation may be user wants to solve the ode by some particular classifier
+    def _finding_algebraic_ode_factor(eqs, func, frac, hard_factor=False):
+
+        from sympy.core.exprtools import factor_terms
+        from sympy.polys.polytools import factor
+
+        if hard_factor:
+            eqs = factor(eqs)
+        eqs = factor_terms(eq, clear=False, fraction=frac)
+        eqs = fraction(eqs)[0] # p/q =0, So we need to solve only p=0
+        eqs_diff = [] # factor containing differential
+        eqs_algebraic = [] # factor containing only algebraic
+        if isinstance(eqs, Mul) and eqs.has(func):
+            fac = eqs.args
+            for i in fac:
+                if i.has(func) and i.has(Derivative):
+                    eqs_diff.append(i)
+                elif i.has(func):
+                    eqs_algebraic.append(i)
+        return eqs_diff, eqs_algebraic
+
+    eqs_diff, eqs_algebraic = _finding_algebraic_ode_factor(eqs, func, frac=True, hard_factor=True) # for factorable type classifier only
+    if len(eqs_diff)+len(eqs_algebraic)>1:
+        r = {'eqs_diff' : eqs_diff,
+            'eqs_algebraic' : eqs_algebraic}
+        matching_hints["factorable"] = r
+
+    eqs_diff, eqs_algebraic = _finding_algebraic_ode_factor(eqs, func, frac=False) # structure of equation will not change
```

Comment:
I don't think it's necessary to split these up. Just send all factors to dsolve. That is why I suggested removing the order==0 check because nth_algebraic would be able to handle the algebraic equations. We should keep this method as simple as possible.

Reply:
This function `_finding_algebraic_ode_factor` will return algebraic and diff equation only if an equation has factors else empty list. And the reason, I am passing equation into this function `_finding_algebraic_ode_factor`  is to remove unnecessary factor(if present). This will help in matching.

eg-
`eq=sqrt(x)*some_equation`
and we are expanding the equation if `classify_ode` so matching is not working. I have found this problem during implementation of Bessel/airy and I find that any 2nd order ode is not able to match and even some of the first order is also not able to match sometime.

Reply:
You should fix that in the other PR. Here we should just get the factors and if they involve the dependent variable then send them back to dsolve. There is no need to distinguish between algebraic and non-algebraic equations since nth_algebraic can handle the algebraic equations (if you remove the order==0 check).